nav2_map_server/src/map_saver_cli_main.cpp

```diff
@@ -0,0 +1,149 @@
+// Copyright 2019 Rover Robotics
+// Copyright (c) 2008, Willow Garage, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include <memory>
+#include <string>
+#include <vector>
+#include <stdexcept>
+
+#include "nav2_map_server/map_mode.hpp"
+#include "nav2_map_server/map_saver.hpp"
+
+#include "rclcpp/rclcpp.hpp"
+
+using namespace nav2_map_server;
+
+const char * USAGE_STRING{
+  "Usage:\n"
+  "  map_saver_cli [arguments] [--ros-args ROS remapping args]\n"
+  "\n"
+  "Arguments:\n"
+  "  -h/--help\n"
+  "  -t <map_topic>\n"
+  "  -f <mapname>\n"
+  "  --occ <threshold_occupied>\n"
+  "  --free <threshold_free>\n"
+  "  --fmt <image_format>\n"
+  "  --mode trinary(default)/scale/raw\n"
+  "\n"
+  "NOTE: --ros-args should be passed at the end of command line"};
+
+
+typedef enum {
+  COMMAND_MAP_TOPIC,
+  COMMAND_MAP_FILE_NAME,
+  COMMAND_IMAGE_FORMAT,
+  COMMAND_OCCUPIED_THRESH,
+  COMMAND_FREE_THRESH,
+  COMMAND_MODE
+} COMMAND_TYPE;
+
+struct cmd_struct
+{
+  const char * cmd;
+  COMMAND_TYPE command_type;
+};
+
+int main(int argc, char ** argv)
+{
+  const struct cmd_struct commands[] = {
+    {"-t", COMMAND_MAP_TOPIC},
+    {"-f", COMMAND_MAP_FILE_NAME},
+    {"--occ", COMMAND_OCCUPIED_THRESH},
+    {"--free", COMMAND_FREE_THRESH},
+    {"--mode", COMMAND_MODE},
+    {"--fmt", COMMAND_IMAGE_FORMAT},
+  };
+
+  rclcpp::init(argc, argv);
+
+  auto logger = rclcpp::get_logger("map_saver_cli");
+
+  std::vector<std::string> arguments(argv + 1, argv + argc);
```

Comment:
Can you move all this arg parsing stuff into an inline method above L74 - 134? that will make this main() clearer to readers:
- init
- parse
- call
- exit

Its easy with 1 big method to loose sight at what its actually doing.

Reply:
Added arguments parsing routine: https://github.com/ros-planning/navigation2/commit/89ece9672dc847161607acc658b63fe2bafb90a7#diff-fe32227e0267593f7bd9f15abfbcc533R69 and some comments to code